### PR TITLE
UOE-10232: AmpMultiformat feature level control

### PR DESCRIPTION
**Pull request by AvinashKapre**

### CODE REVIEW

File: modules/pubmatic/openwrap/publisherfeature/tbf.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			if traffic < 0 || traffic > 100 {
+				profTraffic[profID] = 0
+			}
+		}
+	}
+}
+
+// updateTBFConfigMap updates the TBF configuration maps from the publisher-feature data.
+func (fe *feature) updateTBFConfigMap() {
+	if fe.publisherFeature == nil {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/publisherfeature/tbf.go:36]: can we combine TBF and AMP updation logic.

File: modules/pubmatic/openwrap/publisherfeature/ampmultiformat.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+// updateAmpMutiformatEnabledPublishers updates the ampMultiformat enabled publishers
+func (fe *feature) updateAmpMutiformatEnabledPublishers() {
+	if fe.publisherFeature == nil {
+		return
+	}
+
+	enabledPublishers := make(map[int]struct{})
+	for pubID, feature := range fe.publisherFeature {
+		for featureID, featureDetails := range feature {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/publisherfeature/ampmultiformat.go:19]: no need of loop.

Comment (AvinashKapre) [reply modules/pubmatic/openwrap/publisherfeature/ampmultiformat.go]: done

**Review by ashishshinde-pubm — APPROVED** [review]

File: modules/pubmatic/openwrap/publisherfeature/fullscreenclickability.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	return nil
+}
+
+/*
+IsUnderFSCThreshold:- returns fsc 1/0 based on:
+1. When publisher has disabled FSC in DB, return 0
+2. If FSC is enabled for publisher(default), consider DSP-threshold , and predict value of fsc 0 or 1.
+3. If dspId is not present return 0
+*/
+func (fe *feature) isUnderFSCThreshold(pubid int, dspid int) int {
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/publisherfeature/fullscreenclickability.go:45]: kept the behaviour same as earlier fsc code.

File: modules/pubmatic/openwrap/publisherfeature/ampmultiformat.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+func (fe *feature) IsAmpMultformatEnabled(pubid int) bool {
+	fe.RLock()
+	defer fe.RUnlock()
+
+	if _, isPresent := fe.ampMultiformat.enabledPublishers[pubid]; isPresent {
+		return true
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/publisherfeature/ampmultiformat.go:37]: Done

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 5]
```diff
@@ -578,7 +580,7 @@ func (m *OpenWrap) applyProfileChanges(rctx models.RequestCtx, bidRequest *openr
 
 func (m *OpenWrap) applyVideoAdUnitConfig(rCtx models.RequestCtx, imp *openrtb2.Imp) {
 	//For AMP request, if AmpVideoEnabled is true then crate a empty video object and update with adunitConfigs
-	if rCtx.AmpVideoEnabled {
+	if rCtx.AmpVideoEnabled && m.featureConfig.IsAmpMultiformatEnabled(rCtx.PubID) {
```
Comment (pm-isha-bharti) [inline modules/pubmatic/openwrap/beforevalidationhook.go:583]: rCtx.AmpVideoEnabled =  true means we have publisher feature enabled already
We can remove the additional check: `m.featureConfig.IsAmpMultiformatEnabled(rCtx.PubID)`
ref line: 252

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 6]
```diff
@@ -618,7 +620,7 @@ func (m *OpenWrap) applyVideoAdUnitConfig(rCtx models.RequestCtx, imp *openrtb2.
 	}
 
 	//For AMP request if AmpVideoEnabled is true then, update the imp.video object with adunitConfig and if adunitConfig is not present then update with default values
-	if rCtx.AmpVideoEnabled {
+	if rCtx.AmpVideoEnabled && m.featureConfig.IsAmpMultiformatEnabled(rCtx.PubID) {
```
Comment (pm-isha-bharti) [inline modules/pubmatic/openwrap/beforevalidationhook.go:623]: We can remove the additional check: m.featureConfig.IsAmpMultiformatEnabled(rCtx.PubID)
ref line: 252

File: modules/pubmatic/openwrap/beforevalidationhook.go [hunk 7]
```diff
@@ -247,7 +249,7 @@ func (m OpenWrap) handleBeforeValidationHook(
 				}
 			}
 			videoAdUnitCtx = adunitconfig.UpdateVideoObjectWithAdunitConfig(rCtx, imp, div, payload.BidRequest.Device.ConnectionType)
-			if rCtx.Endpoint == models.EndpointAMP && isVideoEnabledForAMP(videoAdUnitCtx.AppliedSlotAdUnitConfig) {
+			if rCtx.Endpoint == models.EndpointAMP && isVideoEnabledForAMP(videoAdUnitCtx.AppliedSlotAdUnitConfig) && m.featureConfig.IsAmpMultiformatEnabled(rCtx.PubID) {
```
Comment (pm-isha-bharti) [inline modules/pubmatic/openwrap/beforevalidationhook.go:252]: Nitpick: Add feature check before the adunit config

Comment (pm-isha-bharti) [reply modules/pubmatic/openwrap/beforevalidationhook.go:252]: This logic will function for a single impression which is fine as of now with AMP, as it always has a single impression. But if we can fix it easily, we should address it now.

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 5 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/beforevalidationhook.go:583]: Done

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 6 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/beforevalidationhook.go:623]: Done

File: modules/pubmatic/openwrap/beforevalidationhook.go (hunk 7 above)
Comment (AvinashKapre) [reply modules/pubmatic/openwrap/beforevalidationhook.go:252]: This requires imp level ```AmpVideoEnabled``` feild. current amp is working through the hb will create the seperate task for this changes.

**Review by pm-isha-bharti — APPROVED** [review]

**Pull request merged**